Make webhook retry backoff parameters configurable

tkashem opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Currently webhook retry backoff parameters are hard coded, we want to have the ability to configure the backoff parameters for webhook retry logic.  

The objective of this PR is to make the backoff parameters configurable only. So the default that is being used today should still be in effect.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

fedebongio commented:
/assign @deads2k 
/triage accepted

fedebongio commented:
/cc @liggitt @jpbetz @roycaihw

tkashem commented on staging/src/k8s.io/apiserver/pkg/server/options/authorization.go:
```diff
@@ -33,6 +34,15 @@ import (
 
+// DefaultAuthorizationWebhookRetryBackoff is the default backoff parameters for
+// authorization webhook used by the apiserver.
+var DefaultAuthorizationWebhookRetryBackoff = wait.Backoff{
```
Maybe we can collapse both `DefaultAuthorizationWebhookRetryBackoff` and `DefaultAuthenticationWebhookRetryBackoff` into a shared backoff `DefaultWebhookRetryBackoff`.

tkashem commented:
/retest

deads2k commented on pkg/kubeapiserver/authenticator/config.go:
```diff
@@ -68,6 +69,8 @@ type Config struct {
 	WebhookTokenAuthnConfigFile string
 	WebhookTokenAuthnVersion    string
 	WebhookTokenAuthnCacheTTL   time.Duration
+	// WebhookRetryBackoff specifies the backoff parameters of webhook retry
```
maybe tell me why I would want to set this.  Our example use-case could be helpful.

deads2k commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -101,9 +102,10 @@ type TokenFileAuthenticationOptions struct {
+	ConfigFile   string
+	Version      string
+	CacheTTL     time.Duration
+	RetryBackoff wait.Backoff
```
maybe tell me why I would want to set this. Our example use-case could be helpful.

deads2k commented on pkg/kubeapiserver/options/authentication.go:
```diff
@@ -419,6 +425,7 @@ func (o *BuiltInAuthenticationOptions) ToAuthenticationConfig() (kubeauthenticat
 		ret.WebhookTokenAuthnConfigFile = o.WebHook.ConfigFile
 		ret.WebhookTokenAuthnVersion = o.WebHook.Version
 		ret.WebhookTokenAuthnCacheTTL = o.WebHook.CacheTTL
+		ret.WebhookRetryBackoff = o.WebHook.RetryBackoff
```
if we do this, why do we need https://github.com/kubernetes/kubernetes/pull/95705/files#diff-97e779171708fff85d7b1ee3fac01336bf351c39066428e12526d0fd7da5c05cR371 ?

deads2k commented on staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go:
```diff
@@ -54,6 +55,9 @@ type DelegatingAuthenticatorConfig struct {
 
 	RequestHeaderConfig *RequestHeaderConfig
+
+	// WebhookRetryBackoff specifies the backoff parameters of webhook retry
```
Finally found the spot I care about.  maybe tell me why I would want to set this. Our example use-case could be helpful.

Also, I expected to see a `WithCustomRetryBackoff` function like we use for `ResourceBuilderFlags`.

tkashem replied:
@deads2k I added the `WithCustomRetryBackoff` function in `DelegatingAuthenticationOptions` - https://github.com/kubernetes/kubernetes/pull/95705/commits/5df6f2d43697b3076438f46400b3bdfeff0b8d84#diff-249fc1fe96479271dbe3cbd4847642b338cfc33a10f1f033fa606da704462320R213-R217. does that sound correct?

deads2k commented on staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go:
```diff
@@ -54,6 +55,9 @@ type DelegatingAuthenticatorConfig struct {
+	WebhookRetryBackoff wait.Backoff
 }
 
 func (c DelegatingAuthenticatorConfig) New() (authenticator.Request, *spec.SecurityDefinitions, error) {
```
I expected a default backoff wired here.

deads2k commented on staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go:
```diff
@@ -102,49 +112,43 @@ func newGenericWebhook(scheme *runtime.Scheme, codecFactory serializer.CodecFact
 	})
-	return err
+
+	if webhookErr == nil && err == wait.ErrWaitTimeout {
```
pull this change into a separate PR.  it looks good, but it's distinct from the overall change.

deads2k commented on staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go:
```diff
@@ -47,34 +45,34 @@ type tokenReviewer interface {
-func New(kubeConfigFile string, version string, implicitAuds authenticator.Audiences, customDial utilnet.DialFunc) (*WebhookTokenAuthenticator, error) {
-	tokenReview, err := tokenReviewInterfaceFromKubeconfig(kubeConfigFile, version, customDial)
+func New(kubeConfigFile string, version string, implicitAuds authenticator.Audiences, retryBackoff wait.Backoff, customDial utilnet.DialFunc) (*WebhookTokenAuthenticator, error) {
+	tokenReview, err := tokenReviewInterfaceFromKubeconfig(kubeConfigFile, version, webhook.DefaultRetryBackoff(0), customDial)
```
why isn't the passed backoff used here?

deads2k replied:
zero looks like a bug here.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -604,3 +604,32 @@ func poller(interval, timeout time.Duration) WaitFunc {
+
+// ContextBoundExponentialBackoff works with a request context and a Backoff. It ensures that the retry wait never
+// exceeds the deadline specified by the request context.
+func ContextBoundExponentialBackoff(ctx context.Context, backoff Backoff, condition ConditionFunc) error {
```
`ExponentialBackoffWithContext` is more in keeping with the rest of the names.

seh replied:
This would have helped with [this recent search of mine](https://kubernetes.slack.com/archives/C0EG7JC6T/p1601514495044600) (from the "sig-api-machinery" Slack channel).

deads2k commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -604,3 +604,32 @@ func poller(interval, timeout time.Duration) WaitFunc {
+		default:
+		}
+
+		if ok, err := condition(); err != nil || ok {
```
`runConditionWithCrashProtection` I would think

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

deads2k commented on staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go:
```diff
@@ -79,7 +85,13 @@ func (c DelegatingAuthenticatorConfig) New() (authenticator.Request, *spec.Secur
+		// Provide a default if WebhookRetryBackoff has not been set by the user.
+		retryBackoff := c.WebhookRetryBackoff
+		if retryBackoff == nil {
+			retryBackoff = webhooktoken.DefaultRetryBackoff()
```
Is this being left empty somewhere?  I would expect the options to already have defaults set so this is unnecessary.

tkashem replied:
@deads2k yes, somewhere it's left empty. If I don't default here, all the `kind` tests in ci fail. We are throwing error now. So I am hoping I can find the spot.

deads2k commented on staging/src/k8s.io/apiserver/pkg/authorization/authorizerfactory/delegating.go:
```diff
@@ -35,12 +36,24 @@ type DelegatingAuthorizerConfig struct {
 
 func (c DelegatingAuthorizerConfig) New() (authorizer.Authorizer, error) {
+	// Provide a default if WebhookRetryBackoff has not been set by the user.
+	retryBackoff := c.WebhookRetryBackoff
```
I'd rather fail on this and return a clear error.  This is a coding error and most people should be using our recommended options which should have this value set.

tkashem replied:
@deads2k I agree, I fixed it, we throw clear error now. I originally misinterpreted your comment https://github.com/kubernetes/kubernetes/pull/95705#discussion_r511133119 to add defaulting here.

deads2k commented on staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go:
```diff
@@ -36,9 +37,9 @@ const (
+	// DefaultInitialBackoffDelay is the default amount of time to wait before
 	// retrying sending audit events through a webhook.
-	DefaultInitialBackoff = 10 * time.Second
+	DefaultInitialBackoffDelay = 10 * time.Second
```
wow that's long.  Leave it for now, but wow.

deads2k commented:
Please squash what you have and then add the nil check fixes that should be errors in a fresh commit.  This is close.

tkashem committed:
thorw error if webhook retry backoof is not specified

tkashem commented:
/retest

tkashem commented:
/retest

tkashem commented:
/retest

tkashem committed:
make backoff parameters configurable for webhook

Currently webhook retry backoff parameters are hard coded, we want
to have the ability to configure the backoff parameters for webhook
retry logic.

tkashem committed:
set webhook retry backoff parameters for kubelet

deads2k commented:
/lgtm
/approve

sjenning commented:
/approve

sttts commented:
/lgtm
/approve

sftim commented on staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:
```diff
@@ -36,6 +37,17 @@ import (
 	openapicommon "k8s.io/kube-openapi/pkg/common"
 )
 
+// DefaultAuthWebhookRetryBackoff is the default backoff parameters for
+// both authentication and authorization webhook used by the apiserver.
```
If a webhook returns a 5xx response that has `Retry-After` set, the client end (Kubernetes) does not pay attention to that.

Would it be OK to briefly mention in a comment that this implementation does not adjust behavior based on `Retry-After`?
(to be clear - it's OK not to check that header; maybe a future PR can start checking that response header)

dims commented:
/approve
